logrotate: update to freeform

martinetd opened the pull request:
###### Motivation for this change

logrotate is much bigger than intended through its mailutils dependency as reported in #162001

This attempts to resolve the problem by making the dependency only appear when required.

Notes:
 - also included first commit for #161929 that is a release blocker and will (should) get merged first
 - this isn't finished as illustrated by PR being a draft, in particular:
   -  ~setting paths user/group doesn't work and was commented~
   -  the compat renamedoptions from services.logrotate.config to extraConfig was removed as non-working, extraConfig would need to be readded as an explicit option for it to work (not difficult) but it was suggested that extraConfg be removed in https://github.com/NixOS/nixpkgs/issues/162001#issuecomment-1052883067 so I'm not quite sure what to do?
   - coupled with #161929 since the service is no longer immediately started on system switch, it's very likely that a configuration error will slip unnoticed. I'd like to add a validation step (logrotate --debug config) somehow, ideally at build time, but I'm not sure how to do that.
(EDIT: looking around I quite like what nixos/modules/services/networking/bird.nix is doing with a checkPhase on a file that can be disabled through cfg.checkConfig in pkgs.writeTextFile. We can do this easily, I'm mostly done with that)
(EDIT2: done, but doesn't check users -- if someone knows how to do that it'd be great...)
   - some of it is quite ugly and I'll be happy to take any improvement, I've favored compatibility over style but there's definitely ways to make this better...
   - ~EDIT: I have no idea why the manual update check fails -- updating an existing section is not allowed?~

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [x] (Module updates) Added a release notes entry if the change is significant
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

martinetd committed:
logrotate: move mail dependency from package to service

having pkgs.logrotate depend on mailutils brings in quite a bit of dependencies
through mailutil itself and recursive dependency to guile when most people
do not need it.

Remove mailutils dependency from the package, and conditionally add it to the
service if the user specify the mail option either at top level or in a path

Fixes #162001

ju1m commented:
@martinetd regarding the build error for the manual:
> Check DocBook files generated from Markdown are consistent

this is related to a new todo item in the PRs' template:
> - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes

martinetd committed:
nginx/logrotate: run logrotate as nginx user

martinetd commented:
Mostly fixed what I wanted, still need to figure out why the manual check fails now (+ stray tab but that one is easy and isn't related)... out of time for today, later!

Izorkin commented:
Is it possible to add the ability to specify several different files?
```
  services.logrotate = {
    paths = {
      files = [ "file1" "file2" ];
      extraConfig = ''
        daily
      '';
    };
  };
```
Result:
```
file1
file2
{
  daily
}

```

martinetd commented:
> Is it possible to add the ability to specify several different files?

It was meant to be possible yet, I don't think I changed this part and didn't test but this obviously doesn't work as you've pointed out...
I'll add a test and fix it.

martinetd commented:
I'm finishing this up but for your question this actually works, the problem is the usage -- I'm actually curious nix doesn't error on this for you, when I try I get this error as appropriate:
```
error: A definition for option `services.logrotate.paths.extraConfig' is not of type `attribute set of null or boolean or signed integer or strings'. Definition values:
       - In `/home/asmadeus/nixpkgs/nixos/lib/build-vms.nix':
           ''
             daily
           ''
(use '--show-trace' to show detailed location information)
```

There are two errors:
 - paths is an attrset of paths, so you don't define a section directly but need to name it
 - the special attribute key to specify files is 'path', not 'files', this must be kept for compatibility
This appears to work:
```
      services.logrotate.paths = {
        multipath = {
          path = [ "file1" "file2" ];
          #whatever attributes you want
        };
      };
```
And generates the following, which should work (I think):
```
"file1" "file2" {
daily
rotate 20
}
```
(daily and rotate 20 come from default values of frequency and keep;  the current version will also have plenty of empty lines that I've trimmed in an update I'm planning to push as soon as I understand why manual stopped building...)

martinetd commented:
Okay, I think we're good this time. Style aside I see no more big problem.
There are a couple of caveats:
 - the check added for logrotate isn't perfect, logrotate --debug is very picky (users that don't exist will generate an error (I'm stripping users for test), when running as normal user like now su as other users fail (also dealt with by stripping users), files missing when there is no missingok fail the check (enabled by default so not a big problem), if a permission denied happens because an intermediate directory isn't readable to our builder the check will fail (thanksfully there the sandbox is helpful, file will not not exist..))
overall it feels better than no check but it might be surprising to some and probably could deserve some improvement, although I don't see how lest of patching logrotate or writing our own verifier....
 - as said in original message I've favored backwards compatibility over style, there's much to be left desired. I didn't bother checking for 'mail' attribute in extraConfig though so these users will need adjusting, that might be worth a warning...

Happy to adjust things as requested, please have a look @ius @Artturin @aanderse @dasJ ! thanks!

ju1m commented on nixos/modules/services/logging/logrotate.nix:
```diff
@@ -82,27 +83,67 @@ let
+    (mapAttrsToList generateLine settings)
+    ++ [
+        # user and group must be handled together
+        (if (settings.user or null != null && settings.group or null != null)
```
Since `user` and `group` have default values this should be equivalent to `settings.user != null` and `settings.group != null`. If you ever need to see if a module option has been defined, you may add `options` to the `{ config, lib, pkgs, ... }:` and use `options.services.logrotate.settings.foo.isDefined`.

martinetd replied:
Yes and no, that would be true for `paths` attrsets but `generateSection` is also used with `services.logrotate.settings` which doesn't have default values there.

ju1m commented on nixos/modules/services/logging/logrotate.nix:
```diff
@@ -82,27 +83,67 @@ let
+      # logrotate --debug also checks that users specified in config
+      # file exist, but we only have sandboxed users here so brown these
+      # out. according to man page that means su, create and createolddir.
+      # XXX find a way to use real /etc/passwd and /etc/groups
```
I don't think it's possible to use the real `/etc/passwd` and `/etc/groups` at build time, especially when `mutableUsers == true`.

ju1m commented on nixos/tests/logrotate.nix:
```diff
@@ -6,31 +6,94 @@ import ./make-test-python.nix ({ pkgs, ...} : rec {
-  machine = { ... }: {
+  nodes = {
+    # default machine
+    defMachine = { ... }: {
```
`defaultMachine` would be clearer and needs no comment.

ju1m commented on nixos/modules/services/logging/logrotate.nix:
```diff
@@ -36,15 +33,17 @@ let
         default = null;
         description = ''
-          The user account to use for rotation.
+          The user account to use for rotation. This setting is deprecated, please use
```
If that settings is to be deprecated I would use `mkRemovedOptionModule` with a message indicating to use `su`, because a simple change in the description goes easily unnoticed.
And simplify the module and test by removing the new code to support the backward compatibility.

martinetd replied:
I wanted to leave a window with the setting deprecated but still working, it's possible by adding warnings by setting config.warnings conditionally (plenty of examples about) but I didn't get around to it yet.

I agree the description isn't enough in itself, but I think warnings can be added in a subsequent PR so as not to slow this down -- I don't really like PRs growing in scope all the time and prefer to keep goals somewhat constrainted... But if I get around to it fast enough I'll update this PR when I'm done.

martinetd replied:
(that being said, if the consensus is to deprecate settings the hard way, it'll definitely make this module a bit simpler and I don't particularly mind for myself. We can easily remove everywhere I've marked deprecated and most of the exceptions I added to generateConfig)

ju1m commented on nixos/modules/services/logging/logrotate.nix:
```diff
@@ -130,26 +171,46 @@ in
+        default = null;
+        description = ''
+          Extra lines prepended to logrotate configuration file.
+          Deprecated, use settings instead.
```
For a prettier rendering, you may use: `<xref linkend="opt-services.logrotate.settings"/>`

ju1m reviewed (commented):
For a consistent code source style, you may run [nixpkgs-fmt](https://github.com/nix-community/nixpkgs-fmt) on the `.nix` files (in a dedicated commit to facilitate reviewing).
Great work @martined! Thanks for having taken time to revamp this service with so much attention to corner cases. :)

martinetd reviewed (commented):
Thanks for the review! I'll fix the two suggestions I didn't reply to tonight, and check nixpkgs-fmt as well.

Izorkin commented:
@martinetd i am currently writing the logrotate configuration with extraConf. The resulting file looks like this:
```
notifempty
olddir all_log
daily
nocompress
rotate 30
missingok
dateext
dateformat -%Y-%m-%d
nomail
ifempty
sharedscripts

/var/log/sshd/auth.log
{
  postrotate
    systemctl reload syslog-ng
  endscript
}
olddir all_log
createolddir 0755 root root
daily
nocompress
rotate 30
missingok
dateext
dateformat _%Y-%m-%d
dateyesterday
datehourago
nomail
ifempty
sharedscripts

/var/data/web/37101/domains/elven.pw/logs/nginx/elven.pw_*.log
{
  create 0640 web-nginx www-user1
  postrotate
    [ ! -f /run/nginx/nginx.pid ] || kill -USR1 `cat /run/nginx/nginx.pid`
  endscript
}

...
/var/log/nginx/access.log
/var/log/nginx/error.log
/var/data/web/all/logs/nginx/access.log
/var/data/web/all/logs/nginx/error.log
{
  create 0640 web-nginx web-nginx
  postrotate
    [ ! -f /run/nginx/nginx.pid ] || kill -USR1 `cat /run/nginx/nginx.pid`
  endscript
}

```
Instead of this variant:
```
"file1" "file2" {
daily
rotate 20
}
```
It is possible to generate this variant:
```
"file1"
"file2"
{
daily
rotate 20
}
```
Will work too.

Izorkin commented:
Another option would be to add the option of writing all the parameters to the very top of the configuration, which are used by default. The parameters specified below in the file overwrite the upper parameters.

Izorkin commented:
```
  services.logrotate = {
    extraConfigDefault = ''
      daily
      nocompress
      rotate 30
      missingok
      dateext
      dateformat -%Y-%m-%d
      nomail
      ifempty
      sharedscripts
    '';
  };
```

martinetd commented:
@Izorkin
> multiple files for a block
oh, I didn't realize the syntax also allowed multiple files on different lines before the block opening bracket. Since both appear to work do you see a reason we should switch to that, or is the current behaviour good for you?

> Another option would be to add the option of writing all the parameters to the very top of the configuration, which are used by default. The parameters specified below in the file overwrite the upper parameters.

I updated `services.logrotate.extraConfig` to add lines at top of the file instead of at bottom of the file in an earlier PR for this reason ( #159187 ), so you might want to try again. With this PR the new way of doing it is with the `services.logrotate.settings` attrset, which will also appear at the bottom.
`services.logrotate.paths` have a priority attribute to control order in the file which would be interesting to somehow apply to settings instead of just going for the start of the file, but unfortunately I don't see how without making it annoying to use (e.g. making settings an attrset of attrsets with settings + priority); I think it's simpler just going for the top

Izorkin commented:
> oh, I didn't realize the syntax also allowed multiple files on different lines before the block opening bracket. Since both appear to work do you see a reason we should switch to that, or is the current behaviour good for you?

This option is easier to read, for example for debugging.
```
"file1"
"file2"
{
daily
rotate 20
}
```

I will check the config later.

Izorkin commented:
> I updated `services.logrotate.extraConfig` to add lines at top of the file instead of at bottom of the file in an earlier PR for this reason ( #159187 ), so you might want to try again. With this PR the new way of doing it is with the `services.logrotate.settings` attrset, which will also appear at the ~bottom~ top! top I meant top.

Then need to remove this option from the list of deprecated.

martinetd commented:
> Then need to remove this option from the list of deprecated.

No, `services.logrotate.extraConfig` is deprecated in favor of `services.logrotate.settings` as said in the option description. Both achieve the same result.

> `"file1" "file2" {` vs `"file1"\n"file2"\n{`

I find new lines ambiguous, in particular I don't understand how logrotate makes the difference between a file named "dateext" and the option "dateext" in the following config:
```
dateext
file1
{
daily
rotate 1
}
```
(after testing it turns out logrotate handles "file1" as a filenames if it starts with a / or is quoted, the above snippet generates an error)

In practice the generated config will have a new line between global options and quotes, so there will also be a visual clue, but I don't particularly find either more readable than the other.
On a busy debian system, for single files I have 16 inline (`/file {`) and 2 separated by a newline (`/file\n{`), for multiple files I have two of each so there is no clear cut really... I guess that I can agree that a very long list is painful to read, and since I don't care I probably ought to change it... and it then wouldn't make sense to try to keep single items on the same line, so, ok, sold; I'll update this as well later with the rest.

aanderse commented:
It's there ever a situation where order is important in these configuration files?

Izorkin commented:
@aanderse I posted an example config in the post above which I use myself.

martinetd commented:
> It's there ever a situation where order is important in these configuration files?

only for "top level" configuration options, which only apply for blocks after them.
So with the current configuration generation (all settings go first), or previous one (all went last) then order doesn't matter as far as I understand... It would matter if we could order the top level settings to generate something like
```
/files {
  daily
}
dateext
/other_files {
  daily
}
```
then only "/other_files" would be affected by dateext.

As things stand, we cannot write the "dateext" line from configuration, so not much can be done with the priority...

aanderse commented:
I like the idea of structured configuration through a `settings` option and I think you have done a good job of that here. I think we need an escape hatch, though. Clearly the way you have written an escape hatch is causing you some trouble - you have tried very hard to be accommodating and I applaud you for that. In my opinion, though, we should go hard or go home: add a `configFile` option so the user can opt out of everything and provide their own text file for config. The format is simple yet quirky enough that I think this is a good idea. As an added bonus you don't have to feel guilty about keeping backwards compatibility for a release.

Take a look at the [mysql](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/services/databases/mysql.nix) module and how it declare the `configFile` option as an example if you want an example.

What do you think?

aanderse commented:
@martinetd you have thought about this more than I have and I like what you're saying. I'm especially a fan of this comment:
> take more time to define what kind of interface we'd like for the New Logrotate Module.

Given everything you have said it sounds like I will support what you propose as a final solution.

ping @ryantm as a *potentially* interested party.

martinetd commented:
Alright, let's get refocused -- I've opened #162225 for the mail / derivation size problem, we can take our time here :)

Since it looks like you're fine with my suggestion I'll try it out and flesh a prototype so we can compare. There honestly are a lot of subtleties I'm not familiar with so I'm learning as I go and some things I suggested might not be as practical as I think, trying things out works best for me... I've converted this to draft and will update it in a day or two when I've had time to rework this.

martinetd committed:
logrotate: run through nixpkgs-fmt

Running once now will make further patches formatting easier

martinetd committed:
nixos/logrotate: convert to freeform

using freeform is the new standard way of using modules and should replace
extraConfig.
In particular, this will allow us to place a condition on mails

martinetd committed:
logrotate: add services.logrotate.configFile escape hatch

martinetd committed:
logrotate: add configuration check at build time

Now the service no longer starts immediately,
check if the config we generated makes sense as soon as possible.

The check isn't perfect because logrotate --debug wants to check
users required, there are two problems:
 - /etc/passwd and /etc/group are sandboxed and we don't have
visibility of system users
 - the check phase runs as nixbld which cannot su to other users
and logrotate fails on this

Until these two problems can be addressed, users-related checks
are filtered out, it's still much better than no check.
The check can be disabled with services.logrotate.checkConfig
if required
(bird also has a preCheck param, to prepare the environment
before check, but we can add it if it becomes necessary)

Since this makes for very verbose builds, we only show errors:
There is no way to control log level, but logrotate hardcodes
'error:' at common log level, so we can use grep, taking care
to keep error codes

Some manual tests:
───────┬──────────────────────────────────────────
       │ File: valid-config.conf
───────┼──────────────────────────────────────────
   1   │ missingok
───────┴──────────────────────────────────────────
logrotate --debug ok
grep ok

───────┬──────────────────────────────────────────
       │ File: postrotate-no-end.conf
───────┼──────────────────────────────────────────
   1   │ missingok
   2   │ /file {
   3   │    postrotate
   4   │      test
   5   │ }
───────┴──────────────────────────────────────────
error: postrotate-no-end.conf:prerotate, postrotate or preremove without endscript

───────┬──────────────────────────────────────────
       │ File: missing-file.conf
───────┼──────────────────────────────────────────
   1   │ "test" { daily }
───────┴──────────────────────────────────────────
error: stat of test failed: No such file or directory

───────┬──────────────────────────────────────────
       │ File: unknown-option.conf
───────┼──────────────────────────────────────────
   1   │ some syntax error
───────┴──────────────────────────────────────────
logrotate --debug ok
error: unknown-option.conf:1 unknown option 'some' -- ignoring line

───────┬──────────────────────────────────────────
       │ File: unknown-user.conf
───────┼──────────────────────────────────────────
   1   │ su notauser notagroup
───────┴──────────────────────────────────────────
error: unknown-user.conf:1 unknown user 'notauser'

In particular note that logrotate would not error on unknown option
(it just ignores the line) but this change makes the check fail.

martinetd committed:
logrotate: add logrotate-checkconf.service

the build-time check is not safe (e.g. doesn't protect from bad users or nomissingok
paths missing), so add a new unit for configuration switch time check

martinetd commented:
Alright, I think this is getting ready.
 - I've updated all descriptions and release notes, should be clear paths/extraConfig are going away in the future
 - I've also added the logrotate-checkconfig service I was thinking of, so we have:
   - basic syntax errors caughts at build time (which I left because it's better to catch what we can early)
   - environment-dependant errors (e.g. trying to use a non-existing users, or nomissingok files missing) caught on configuration switch:
```
# /nix/store/rbjhsv675qjwbgiprj35zih6mq0r8xfd-nixos-system-failingMachine-22.05pre-git/bin/switch-to-configuration test
starting the following units: logrotate-checkconf.service, nscd.service
Job for logrotate-checkconf.service failed because the control process exited with error code.
See "systemctl status logrotate-checkconf.service" and "journalctl -xeu logrotate-checkconf.service" for details.
warning: the following units failed: logrotate-checkconf.service

× logrotate-checkconf.service - Logrotate configuration check
     Loaded: loaded (/etc/systemd/system/logrotate-checkconf.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Sat 2022-03-05 01:21:06 UTC; 183ms ago
    Process: 1056 ExecStart=/nix/store/bifjan2fvp6k9p3jk4fyv1bykynsj623-logrotate-3.19.0/sbin/logrotate --debug /nix/store/dhd40ayh0wpqzzr6rwgmpdavfq7awr7w-logrotate.conf (code=exited, status=1/FAILURE)
   Main PID: 1056 (code=exited, status=1/FAILURE)
         IP: 0B in, 0B out
        CPU: 2ms

Mar 05 01:21:06 failingMachine logrotate[1056]: error: /nix/store/dhd40ayh0wpqzzr6rwgmpdavfq7awr7w-logrotate.conf:2 unknown user 'notarealuser'
Mar 05 01:21:06 failingMachine logrotate[1056]: Reading state from file: /var/lib/logrotate.status
Mar 05 01:21:06 failingMachine logrotate[1056]: state file /var/lib/logrotate.status does not exist
Mar 05 01:21:06 failingMachine logrotate[1056]: Allocating hash table for state file, size 64 entries
Mar 05 01:21:06 failingMachine logrotate[1056]: Handling 0 logs
Mar 05 01:21:06 failingMachine systemd[1]: Failed to start Logrotate configuration check.
Mar 05 01:21:06 failingMachine systemd[1]: Starting Logrotate configuration check...
Mar 05 01:21:06 failingMachine systemd[1]: logrotate-checkconf.service: Main process exited, code=exited, status=1/FAILURE
Mar 05 01:21:06 failingMachine systemd[1]: logrotate-checkconf.service: Failed with result 'exit-code'.
Mar 05 01:21:06 failingMachine systemd[1]: Failed to start Logrotate configuration check.
```

The rest hasn't changed from my message two days ago. There's no hurry, please take your time and test/comment :)

ju1m reviewed (approved):
LGTM

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/757

Izorkin commented:
Small fix:
```diff
diff --git a/nixos/modules/system/boot/systemd.nix b/nixos/modules/system/boot/systemd.nix
index 926d9a95843..fd0684437db 100644
--- a/nixos/modules/system/boot/systemd.nix
+++ b/nixos/modules/system/boot/systemd.nix
@@ -1219,7 +1219,7 @@ in
     boot.kernelParams = optional (!cfg.enableUnifiedCgroupHierarchy) "systemd.unified_cgroup_hierarchy=0";

     services.logrotate.settings = {
-      "/ar/log/btmp" = mapAttrs (_: mkDefault) {
+      "/var/log/btmp" = mapAttrs (_: mkDefault) {
         frequency = "monthly";
         rotate = 1;
         create = "0660 root ${config.users.groups.utmp.name}";

```

martinetd commented:
> Small fix:
> /ar/log -> /var/log

ugh! good catch, I've spent too long on these patches... So much for tests only looking at wtmp! Force pushing a fix.

Izorkin commented:
If use `nixops` error check` logrotate.conf`. The `nginx service` is not installed on this server.
```
error: /tmp/logrotate.conf:36 unknown user 'nginx'
error: found error in "/var/log/nginx/access.log"
error: builder for '/nix/store/byp1s9wd1k1gmd8hyklz03s6a9aram99-logrotate.conf.drv' failed with exit code 1;
       last 2 log lines:
       > error: /tmp/logrotate.conf:36 unknown user 'nginx'
       > error: found error in "/var/log/nginx/access.log"
       For full logs, run 'nix log /nix/store/byp1s9wd1k1gmd8hyklz03s6a9aram99-logrotate.conf.drv'.

```

Izorkin commented:
Fix variant - add option `services.logrotate.diableCheck = true`.

martinetd commented:
> If use nixops error check logrotate.conf. The nginx service is not installed on this server.

hm, how does that happen?
The build check will fail if there is a rule specifying users, but it should sed su/create/createolddir rules out to avoid the problem, and I don't see how you'd have something with logrotate if the service isn't enabled...

Could you print the content of /tmp/logrotate.conf  in configFile's checkPhase ?
e.g.
```diff
diff --git a/nixos/modules/services/logging/logrotate.nix b/nixos/modules/services/logging/logrotate.nix
index e4fd5ec1ae8e..29916f14e575 100644
--- a/nixos/modules/services/logging/logrotate.nix
+++ b/nixos/modules/services/logging/logrotate.nix
@@ -164,6 +164,7 @@ let
           -e "s/\(create \+[0-9]* *\|createolddir \+[0-9]* \+\) .*/\1 $user $group/" \
           -e "1imissingok" -e "s/nomissingok//" \
           $out > /tmp/logrotate.conf
+      cat /tmp/logrotate.conf
       # Since this makes for very verbose builds only show real error.
       # There is no way to control log level, but logrotate hardcodes
       # 'error:' at common log level, so we can use grep, taking care

```

Yes check can be disabled but we shouldn't fail so easily, I'd like to understand and fix it. Probably need to expand the sed replace rule and add whatever you have on line 36 there..

Izorkin commented:
Config:
```
  services.logrotate = {
    settings = {
      nginx = lib.mkIf config.services.nginx.enable {
        olddir = "old_logs";
        createolddir = "0750 ${config.services.nginx.user} ${config.services.nginx.group}";
      };
    };
  };

```

Error:
```
"/var/log/nginx/*.log" {
  compress
  createolddir 0750 nginx nginx
  delaycompress
  weekly
  olddir old_logs
  postrotate
    [ ! -f /var/run/nginx/nginx.pid ] || kill -USR1 `cat /var/run/nginx/nginx.pid`
  endscript
  rotate 26
  su nixbld nixbld
}
error: /tmp/logrotate.conf:36 unknown user 'nginx'
error: found error in "/var/log/nginx/*.log" , skipping
error: builder for '/nix/store/ggv0ailjx6qbf9phsq99p5aikr3vin7f-logrotate.conf.drv' failed with exit code 1;
       last 10 log lines:
       >   weekly
       >   olddir old_logs
       >   postrotate
       >     [ ! -f /var/run/nginx/nginx.pid ] || kill -USR1 `cat /var/run/nginx/nginx.pid`
       >   endscript
       >   rotate 26
       >   su nixbld nixbld
       > }
       > error: /tmp/logrotate.conf:36 unknown user 'nginx'
       > error: found error in "/var/log/nginx/*.log" , skipping
       For full logs, run 'nix log /nix/store/ggv0ailjx6qbf9phsq99p5aikr3vin7f-logrotate.conf.drv'.

```

Izorkin commented:
I ise this logrotate config for nginx:
```
```
/var/log/nginx/site1_*.log
{
  create 0640 web-nginx www-user1
  postrotate
    [ ! -f /run/nginx/nginx.pid ] || kill -USR1 `cat /run/nginx/nginx.pid`
  endscript
}

/var/log/nginx/site2_*.log
{
  create 0640 web-nginx www-user2
  postrotate
    [ ! -f /run/nginx/nginx.pid ] || kill -USR1 `cat /run/nginx/nginx.pid`
  endscript
}
```
```

martinetd commented:
> createolddir 0750 nginx nginx

Thanks! I screwed up the regex indeed...

>           -e "s/\(create \+[0-9]* *\|createolddir \+[0-9]* \+\) .*/\1 $user $group/" \
requires two spaces after mode to match (one for " \+" and one before .*)

Could you try the following replace?
```
      sed -e "s/\bsu\s.*/su $user $group/" \
          -e "s/\b\(create\s\+[0-9]*\s*\|createolddir\s\+[0-9]*\s\+\).*/\1$user $group/" \
          -e "1imissingok" -e "s/\bnomissingok\b//" \
          $out > /tmp/logrotate.conf
```

I've made it more robust all around: fixed the double-space requirement, and added anchors (the 'su' replace rule was a bit too broad in my opinion)

I'd also like to update the test so we print logrotate.conf in case of failure (so user can debug it more easily) but don't have time to update the patch right now, will push a fix tomorrow morning.

martinetd commented:
> I'd also like to update the test so we print logrotate.conf in case of failure (so user can debug it more easily) but don't have time to update the patch right now, will push a fix tomorrow morning.

I lied! :)

Here's what an error looks like now:
```
logrotate.conf> Logrotate configuration check failed.
logrotate.conf> The failing configuration (after adjustments to pass tests in sandbox) was:
logrotate.conf> -------
logrotate.conf> missingok
logrotate.conf> delaycompress
....
logrotate.conf> -------
logrotate.conf> The error reported by logrotate was as follow:
logrotate.conf> -------
logrotate.conf> error: /tmp/logrotate.conf:23 unknown option 'createdir' -- ignoring line
logrotate.conf> error: /tmp/logrotate.conf:24 unknown option 'createdir' -- ignoring line
logrotate.conf> -------
logrotate.conf> You can disable this check with services.logrotate.checkConfig = false,
logrotate.conf> but if you think it should work please report this failure along with
logrotate.conf> the config file being tested!
error: builder for '/nix/store/9vgk3qg0zc33pbzps09k7bhbdkpn8hip-logrotate.conf.drv' failed with exit code 1;
       last 10 log lines:
       > shred
       > -------
       > The error reported by logrotate was as follow:
       > -------
       > error: /tmp/logrotate.conf:23 unknown option 'createdir' -- ignoring line
       > error: /tmp/logrotate.conf:24 unknown option 'createdir' -- ignoring line
       > -------
       > You can disable this check with services.logrotate.checkConfig = false,
       > but if you think it should work please report this failure along with
       > the config file being tested!
       For full logs, run 'nix log /nix/store/9vgk3qg0zc33pbzps09k7bhbdkpn8hip-logrotate.conf.drv'.
```

I've also added a check for createolddir/createdir replacements (su already was tested in the legacy section, but as that will be removed I've added it as well).

This is all in a new patch to be easier to review for anyone who already did review the code.

Thanks for taking the time to test!

Izorkin commented:
@martinetd thanks, working!

Izorkin commented:
Maybe add these to defaults?
```
  services.logrotate = {
    settings = {
      header = {
        ...
        dateyesterday = true;
        datehourago = true;
      };

```

Izorkin commented:
I agree with this minimal variant:
```
  services.logrotate = {
    settings = {
      header = {
        rotate = "4";
        frequency = "weekly";
      };
```

martinetd commented:
I've updated just frequencies as discussed above. Note this doesn't change any existing (obsolete) paths default, as the paths suboption attrs defaults will prevail, so it's pretty safe to do now.
Leaving the changed from notifempty to create for later unless others think we should update our defaults to be closer from other distros.

Izorkin commented:
@martinetd dont`t needed remove `missingok` and `notifempty` from `header`?

martinetd commented:
Ah, I thought you meant the comment as adding the two -- if we remove missingok we definitely want create, as otherwise a stopped service would get logrotate stuck. logrotates defaults here don't make sense.

If we also add create I don't mind removing the other two, but these are long-standing nixos defaults so might break some setup unlike the other two.

martinetd commented:
Sorry -- I think I prefer to err on the safety side with the defaults, this PR in itself shouldn't break anything so let's leave it as is unless others feel strongly about it. We can leave the missingok/notifempty -> create change to a later PR if someone thinks it's important

@aanderse any comment on the new version? What schedule would you want to move forward? 20.05 is getting close, so would be good to merge it this month to leave some room for adjustments when things will go wrong

aanderse commented:
@martinetd honest opinion? I think this code is too complex for me to review late at night, which is where most of my time for this type of review comes from...

Looks cool, but I'd be less than honest if I didn't say I think it looks more complex than I think it should. This is mostly attributed to your monumental effort to keep backwards compatibility, though. For a module which is hardly used at all (**but should be**) is the increased code complexity worth it? :man_shrugging:

Don't let me stand in the way of anyone else who has properly reviewed clicking merge though. And to be clear - I appreciate your efforts on this. Thank you. :bow:

martinetd commented:
If you want to review without the backwards compatibility, you can check https://github.com/martinetd/nixpkgs/tree/logrotate_post_cleanup out -- it's the very same branch with one extra commit that removes all the cruft, so the final file can be easier to review.

I'm all for deprecating old API but there should be at least one release in between, e.g. release 22.05 with compat and then rip it out in half a year after people have had time to see the warnings when they upgraded. I think upgrades should be as painless as possible to get it over with, and give time to fix warnings later when there's less going on all at once.


It's still quite complex frankly -- the config file syntax is a mess in the first place so we can't use any common generator, and ordering being semantically important requires some hoops with priority, but it's probably a bit better than the current state... :|

Artturin commented on nixos/modules/services/networking/lxd-image-server.nix:
```diff
@@ -56,11 +56,9 @@ in
         frequency = "daily";
         rotate = 21;
         create = "755 lxd-image-server ${cfg.group}";
-        missingok = true;
```
please put changes to other modules in a separate pr

martinetd replied:
This is a default value, so this doesn't change anything -- it just reflects defaults that had always been there.
As there is compatibility I can drop all the revamp though and open a later PR if you prefer, but folks using it will get a wall of warnings until it's converted. What would you like?

Artturin replied:
nevermind, i was thinking about some other PR while looking at the files

Artturin commented:
add a comment with a link to this pr and that the deprecated options should be removed before 22.11

```patch
From a771eb18ef7f315a8bdbc4b9c2524da4094d1101 Mon Sep 17 00:00:00 2001
From: Dominique Martinet <asmadeus@codewreck.org>
Date: Wed, 2 Mar 2022 22:32:48 +0900
Subject: [PATCH] logrotate service: cleanup deprecated options

---
 nixos/modules/services/logging/logrotate.nix | 181 +------------------
 nixos/modules/services/misc/gitlab.nix       |  11 +-
 nixos/tests/logrotate.nix                    |  29 ---
 3 files changed, 2 insertions(+), 219 deletions(-)

diff --git a/nixos/modules/services/logging/logrotate.nix b/nixos/modules/services/logging/logrotate.nix
index 4550e384c2420..03d3b2604a9b4 100644
--- a/nixos/modules/services/logging/logrotate.nix
+++ b/nixos/modules/services/logging/logrotate.nix
@@ -5,91 +5,9 @@ with lib;
 let
   cfg = config.services.logrotate;
 
-  # deprecated legacy compat settings
-  pathOpts = { name, ... }: {
-    options = {
-      enable = mkOption {
-        type = types.bool;
-        default = true;
-        description = ''
-          Whether to enable log rotation for this path. This can be used to explicitly disable
-          logging that has been configured by NixOS.
-        '';
-      };
-
-      name = mkOption {
-        type = types.str;
-        internal = true;
-      };
-
-      path = mkOption {
-        type = with types; either str (listOf str);
-        default = name;
-        defaultText = "attribute name";
-        description = ''
-          The path to log files to be rotated.
-          Spaces are allowed and normal shell quoting rules apply,
-          with ', ", and \ characters supported.
-        '';
-      };
-
-      user = mkOption {
-        type = with types; nullOr str;
-        default = null;
-        description = ''
-          The user account to use for rotation.
-        '';
-      };
-
-      group = mkOption {
-        type = with types; nullOr str;
-        default = null;
-        description = ''
-          The group to use for rotation.
-        '';
-      };
-
-      frequency = mkOption {
-        type = types.enum [ "hourly" "daily" "weekly" "monthly" "yearly" ];
-        default = "daily";
-        description = ''
-          How often to rotate the logs.
-        '';
-      };
-
-      keep = mkOption {
-        type = types.int;
-        default = 20;
-        description = ''
-          How many rotations to keep.
-        '';
-      };
-
-      extraConfig = mkOption {
-        type = types.lines;
-        default = "";
-        description = ''
-          Extra logrotate config options for this path. Refer to
-          <link xlink:href="https://linux.die.net/man/8/logrotate"/> for details.
-        '';
-      };
-
-      priority = mkOption {
-        type = types.int;
-        default = 1000;
-        description = ''
-          Order of this logrotate block in relation to the others. The semantics are
-          the same as with `lib.mkOrder`. Smaller values have a greater priority.
-        '';
-      };
-    };
-
-    config.name = name;
-  };
-
   generateLine = n: v:
     if builtins.elem n [ "files" "priority" "enable" "global" ] || v == null then null
-    else if builtins.elem n [ "extraConfig" "frequency" ] then "${v}\n"
+    else if builtins.elem n [ "frequency" ] then "${v}\n"
     else if builtins.elem n [ "firstaction" "lastaction" "prerotate" "postrotate" "preremove" ]
          then "${n}\n    ${v}\n  endscript\n"
     else if isInt v then "${n} ${toString v}\n"
@@ -108,25 +26,6 @@ let
         ${generateSection 2 settings}}
     '';
 
-  # below two mapPaths are compat functions
-  mapPathOptToSetting = n: v:
-    if n == "keep" then nameValuePair "rotate" v
-    else if n == "path" then nameValuePair "files" v
-    else nameValuePair n v;
-
-  mapPathsToSettings = path: pathOpts:
-    nameValuePair path (
-      filterAttrs (n: v: ! builtins.elem n [ "user" "group" "name" ] && v != "") (
-        (mapAttrs' mapPathOptToSetting pathOpts) //
-        {
-          su =
-            if pathOpts.user != null
-            then "${pathOpts.user} ${pathOpts.group}"
-            else null;
-        }
-      )
-    );
-
   settings = sortProperties (attrValues (filterAttrs (_: settings: settings.enable) (
     foldAttrs recursiveUpdate { } [
       {
@@ -137,15 +36,7 @@ let
           frequency = "weekly";
           rotate = 4;
         };
-        # compat section
-        extraConfig = {
-          enable = (cfg.extraConfig != "");
-          global = true;
-          extraConfig = cfg.extraConfig;
-          priority = 101;
-        };
       }
-      (mapAttrs' mapPathsToSettings cfg.paths)
       cfg.settings
       { header = { global = true; priority = 100; }; }
     ]
@@ -196,10 +87,6 @@ let
     else "";
 in
 {
-  imports = [
-    (mkRenamedOptionModule [ "services" "logrotate" "config" ] [ "services" "logrotate" "extraConfig" ])
-  ];
-
   options = {
     services.logrotate = {
       enable = mkEnableOption "the logrotate systemd service" // {
@@ -309,76 +196,10 @@ in
           in this case you can disable the failing check with this option.
         '';
       };
-
-      # deprecated legacy compat settings
-      paths = mkOption {
-        type = with types; attrsOf (submodule pathOpts);
-        default = { };
-        description = ''
-          Attribute set of paths to rotate. The order each block appears in the generated configuration file
-          can be controlled by the <link linkend="opt-services.logrotate.paths._name_.priority">priority</link> option
-          using the same semantics as `lib.mkOrder`. Smaller values have a greater priority.
-          This setting has been deprecated in favor of <link linkend="opt-services.logrotate.settings">logrotate settings</link>.
-        '';
-        example = literalExpression ''
-          {
-            httpd = {
-              path = "/var/log/httpd/*.log";
-              user = config.services.httpd.user;
-              group = config.services.httpd.group;
-              keep = 7;
-            };
-
-            myapp = {
-              path = "/var/log/myapp/*.log";
-              user = "myuser";
-              group = "mygroup";
-              frequency = "weekly";
-              keep = 5;
-              priority = 1;
-            };
-          }
-        '';
-      };
-
-      extraConfig = mkOption {
-        default = "";
-        type = types.lines;
-        description = ''
-          Extra contents to append to the logrotate configuration file. Refer to
-          <link xlink:href="https://linux.die.net/man/8/logrotate"/> for details.
-          This setting has been deprecated in favor of
-          <link linkend="opt-services.logrotate.settings">logrotate settings</link>.
-        '';
-      };
     };
   };
 
   config = mkIf cfg.enable {
-    assertions =
-      mapAttrsToList
-        (name: pathOpts:
-          {
-            assertion = (pathOpts.user != null) == (pathOpts.group != null);
-            message = ''
-              If either of `services.logrotate.paths.${name}.user` or `services.logrotate.paths.${name}.group` are specified then *both* must be specified.
-            '';
-          })
-        cfg.paths;
-
-    warnings =
-      (mapAttrsToList
-        (name: pathOpts: ''
-          Using config.services.logrotate.paths.${name} is deprecated and will become unsupported in a future release.
-          Please use services.logrotate.settings instead.
-        '')
-        cfg.paths
-      ) ++
-      (optional (cfg.extraConfig != "") ''
-        Using config.services.logrotate.extraConfig is deprecated and will become unsupported in a future release.
-        Please use services.logrotate.settings with globals=true instead.
-      '');
-
     systemd.services.logrotate = {
       description = "Logrotate Service";
       startAt = "hourly";
diff --git a/nixos/modules/services/misc/gitlab.nix b/nixos/modules/services/misc/gitlab.nix
index 488c3be7b653a..4a4f0c94e7a41 100644
--- a/nixos/modules/services/misc/gitlab.nix
+++ b/nixos/modules/services/misc/gitlab.nix
@@ -225,6 +225,7 @@ in {
     (mkRenamedOptionModule [ "services" "gitlab" "stateDir" ] [ "services" "gitlab" "statePath" ])
     (mkRenamedOptionModule [ "services" "gitlab" "backupPath" ] [ "services" "gitlab" "backup" "path" ])
     (mkRemovedOptionModule [ "services" "gitlab" "satelliteDir" ] "")
+    (mkRemovedOptionModule [ "services" "gitlab" "logrotate" "extraConfig" ] "Modify services.logrotate.settings.gitlab directly instead")
   ];
 
   options = {
@@ -845,15 +846,6 @@ in {
           default = 30;
           description = "How many rotations to keep.";
         };
-
-        extraConfig = mkOption {
-          type = types.lines;
-          default = "";
-          description = ''
-            Extra logrotate config options for this path. Refer to
-            <link xlink:href="https://linux.die.net/man/8/logrotate"/> for details.
-          '';
-        };
       };
 
       extraConfig = mkOption {
@@ -982,7 +974,6 @@ in {
           rotate = cfg.logrotate.keep;
           copytruncate = true;
           compress = true;
-          extraConfig = cfg.logrotate.extraConfig;
         };
       };
     };
diff --git a/nixos/tests/logrotate.nix b/nixos/tests/logrotate.nix
index b0685f3af9ff1..94f6ad5103fb2 100644
--- a/nixos/tests/logrotate.nix
+++ b/nixos/tests/logrotate.nix
@@ -64,29 +64,6 @@ import ./make-test-python.nix ({ pkgs, ... }: rec {
           notifempty = true;
         };
       };
-      # extraConfig compatibility - should be added to top level, early.
-      services.logrotate.extraConfig = ''
-        nomail
-      '';
-      # paths compatibility
-      services.logrotate.paths = {
-        compat_path = {
-          path = "compat_test_path";
-        };
-        # user/group should be grouped as 'su user group'
-        compat_user = {
-          user = config.users.users.root.name;
-          group = "root";
-        };
-        # extraConfig in path should be added to block
-        compat_extraConfig = {
-          extraConfig = "dateext";
-        };
-        # keep -> rotate
-        compat_keep = {
-          keep = 1;
-        };
-      };
     };
   };
 
@@ -127,12 +104,6 @@ import ./make-test-python.nix ({ pkgs, ... }: rec {
               "sed -ne '/\"postrotate\" {/,/}/p' /tmp/logrotate.conf | grep endscript",
               "grep '\"file1\"\n\"file2\" {' /tmp/logrotate.conf",
               "sed -ne '/\"import\" {/,/}/p' /tmp/logrotate.conf | grep noolddir",
-              "sed -ne '1,/^\"/p' /tmp/logrotate.conf | grep nomail",
-              "grep '\"compat_test_path\" {' /tmp/logrotate.conf",
-              "sed -ne '/\"compat_user\" {/,/}/p' /tmp/logrotate.conf | grep 'su root root'",
-              "sed -ne '/\"compat_extraConfig\" {/,/}/p' /tmp/logrotate.conf | grep dateext",
-              "[[ $(sed -ne '/\"compat_keep\" {/,/}/p' /tmp/logrotate.conf | grep -w rotate) = \"  rotate 1\" ]]",
-              "! sed -ne '/\"compat_keep\" {/,/}/p' /tmp/logrotate.conf | grep -w keep",
           )
           # also check configFile option
           failingMachine.succeed(
```

martinetd commented on nixos/modules/services/logging/logrotate.nix:
```diff
@@ -112,17 +203,122 @@ in
+
+          Conversely there are still things that might make this check fail incorrectly
+          (e.g. a file path where we don't have access to intermediate directories, which
+          are quire unlikely (/var does not exist), but possible):
```
/var doesn't exist here is not meant as an example of failure (we force missingok for the build-time test config), but as a way of saying permission issues are unlikely: since /var doesn't exist, anything below /var won't fail with a permission error.
That being said I agree it's confusing, I'll just remove this bit.

martinetd commented:
> add a comment with a link to this pr and that the deprecated options should be removed before 22.11

Thanks for your review!

I've opened a PR with option removal (as draft) and added a comment with a link to it.

Artturin commented on nixos/modules/services/logging/logrotate.nix:
```diff
@@ -6,6 +6,8 @@ let
 
   # deprecated legacy compat settings
+  # these options will be removed before 22.11 in the following PR:
+  # https://github.com/NixOS/nixpkgs/pull/164169k
```
there's a k

Artturin replied:
should be https://github.com/NixOS/nixpkgs/pull/164169

martinetd replied:
ugh, sorry - fixed in a force push -_-

Artturin commented on nixos/modules/services/logging/logrotate.nix:
```diff
@@ -172,7 +387,16 @@ in
+        ExecStart = "${pkgs.logrotate}/sbin/logrotate ${mailOption} ${cfg.configFile}";
+      };
+    };
+    systemd.services.logrotate-checkconf = {
```
it would be better to move this to the `preStart` of the logrotate service

https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStartPre=

> ExecStart= commands are only run after all ExecStartPre= commands that were not prefixed with a "-" exit successfully.

martinetd replied:
I don't understand services that explicitly check their config in prestart. Start itself will do that very same check just fine, I don't see the point of moving this to preStart (well, some services could start in degraded mode with minor errors, but if so for something like logrotate I'd say we want what can be rotated rotated? That's why I didn't add a dependency on this even if this from the main service)

The reason I made this service was that I wanted a warning on configuration switch when the configuration is bad -- because logrotate.service itself was disabled in #161929 as it broke testSwitch, so we've lost this feature recently. I find it more useful to get an immediate feedback in "the following services failed to start" block of config activation than waiting 1h for the next logrotate run and notice it didn't work later (who knows when) 


If you think that's overdoing it we can just remove this extra service, though. Or maybe figure a way to re-enable the main service without making testSwitch fail...

martinetd replied:
(Anyway, that's unrelated to the main goal of this PR, so I'd be happy to just remove it for now and discuss it elsewhere if it's debatable)

Artturin commented on nixos/modules/virtualisation/azure-agent.nix:
```diff
@@ -146,15 +146,11 @@ in
-        }
-      '';
+      settings."/var/log/waagent.log" = {
+        compress = true;
```
could this be a default?

martinetd replied:
compress breaks some logs without delaycompress (if e.g. an application keeps writing to the just-rotated logs for a bit and we compress it we'll lose the last lines), so I think it makes more sense left as is.
If we also add delaycompress (would compress .2 files onwards) then I think it probably is ok to enable by default, but no other distro does this (see https://github.com/NixOS/nixpkgs/pull/162063#issuecomment-1060552102 for comparison between distros), so I'm a bit split on this.

Artturin reviewed (approved):
Seems fine

Built a vm and compared the new generated file to the old one and it was ok

Artturin commented:
squashed locally to preserve commit messages

martinetd commented:
Thanks! I prefer to keep things split slightly though to illustrate logical changes, I've fixed the history up so all fixes commits are squashed with their appropriate counterpart, and we can just e.g. revert the logrotate-checkconf service  commit if this becomes needed later, or simply for easier history reading while keeping the system bisectable (well, at least I've checkede that nixosTests.logrotate passes for each commit individually, so logrotate is "instantiable" at all time)

You can check this results in no code change:
```
$ git diff 4a466cb 1647fd2
<empty>
```

I appreciate your effort of manually squashing though, and have kept your squashed commit around in a local branch if you strongly prefer a single commit for merging.

martinetd commented:
aa0f27abb06ca66a1 introduced a small conflict so I've rebased the branch to today's master.

@aanderse my suggestion to review the final form in #164169 instead still stands if you
can find a moment (should be less convoluted than this as compatibility has been removed),
otherwise I think we're getting stuck -- this has had two proper reviews (from ju1m and
Artturin), should I try to find one more?

Artturin commented:
@ofborg test logrotate gitlab nginx systemd

You have written tests so if these pass I think ill merge

martinetd commented:
> You have written tests so if these pass I think ill merge

The tests passed! thanks!

Artturin commented:
Thanks for your work

aanderse commented:
🚀

wizeman commented on nixos/modules/services/logging/logrotate.nix:
```diff
@@ -86,23 +89,113 @@ let
+  };
+
+  mailOption =
+    if foldr (n: a: a || n ? mail) false (attrValues cfg.settings)
```
Sorry for the late comment, but this doesn't seem to be accounting for when the user sets `mail = false` (so that `nomail` appears in the appropriate section in `logrotate.conf`).

In other words, if the user sets `mail = false` in one of the sections in the settings, the `--mail` option will get added even though no section in the settings actually sets `mail = true`.

martinetd replied:
good point, I'll update it to `a || n.mail or false` in a new PR

martinetd replied:
Done in #177106

wizeman commented on nixos/modules/services/logging/logrotate.nix:
```diff
@@ -86,23 +89,113 @@ let
+
+  mailOption =
+    if foldr (n: a: a || n ? mail) false (attrValues cfg.settings)
+    then "--mail=${pkgs.mailutils}/bin/mail"
```
Again, sorry for the late comment, but shouldn't this be something like:
```Nix
if /* ... */
then "\"--mail=${pkgs.mailutils}/bin/mail -s\""
else ""
```
i.e., shouldn't the `-s` argument be specified?

The `logrotate` man page says (about the `--mail` option):
> This command should accept two arguments: 1) the subject of the message, and 2) the recipient.
> (...)
> The default mail command is `/bin/mail -s`.

martinetd replied:
hm, this comes straight from what we had in the package, from the recent diff:
```
-  ] ++ lib.optionals (mailutils != null) [
-    "--with-default-mail-command=${mailutils}/bin/mail"
   ];
```
So if this doesn't work then it's likely sending mails from logrotate never worked...

I'll admit I never tested either myself, I'll give it a try this weekend unless someone beats me to it.

martinetd replied:
from logrotate changelog:
```
106:  - remove `-s` from `DEFAULT_MAIL_COMMAND` and improve its documentation (#152)
```
You might be looking at old man page? that dates 3.13.0 (2017)....

looking at the code and not testing I think we're fine, my setup is half broken so it's a pain to test and I'll leave it at code inspection...

wizeman replied:
Yes, I think you are correct.
I was looking at `https://linux.die.net/man/8/logrotate`, which is what `man configuration.nix` pointed me at, but it seems that this man page is not up-to-date.
